components/script/dom/notification.rs

```diff
@@ -0,0 +1,444 @@
+};
+use crate::dom::bindings::codegen::UnionTypes::UnsignedLongOrUnsignedLongSequence;
+use crate::dom::bindings::error::Error;
+use crate::dom::bindings::import::module::{Fallible, Rc};
+use crate::dom::bindings::inheritance::Castable;
+use crate::dom::bindings::reflector::reflect_dom_object_with_proto;
+use crate::dom::bindings::root::{Dom, DomRoot};
+use crate::dom::bindings::str::{DOMString, USVString};
+use crate::dom::bindings::trace::RootedTraceableBox;
+use crate::dom::bindings::utils::to_frozen_array;
+use crate::dom::eventtarget::EventTarget;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::permissions::{get_descriptor_permission_state, PermissionAlgorithm, Permissions};
+use crate::dom::promise::Promise;
+use crate::dom::serviceworkerglobalscope::ServiceWorkerGlobalScope;
+use crate::dom::serviceworkerregistration::ServiceWorkerRegistration;
+use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
+
+// TODO: Service Worker API (persistent notification)
+// https://notifications.spec.whatwg.org/#service-worker-api
+
+/// <https://notifications.spec.whatwg.org/#notifications>
+#[dom_struct]
+pub(crate) struct Notification {
+    eventtarget: EventTarget,
+    /// <https://notifications.spec.whatwg.org/#service-worker-registration>
+    serviceworker_registration: Option<Dom<ServiceWorkerRegistration>>,
+    /// <https://notifications.spec.whatwg.org/#concept-title>
+    title: DOMString,
+    /// <https://notifications.spec.whatwg.org/#body>
+    body: DOMString,
+    /// <https://notifications.spec.whatwg.org/#data>
+    #[ignore_malloc_size_of = "mozjs"]
+    data: Box<Heap<JSVal>>,
+    /// <https://notifications.spec.whatwg.org/#concept-direction>
+    dir: NotificationDirection,
+    /// <https://notifications.spec.whatwg.org/#image-url>
+    image: Option<USVString>,
+    /// <https://notifications.spec.whatwg.org/#icon-url>
+    icon: Option<USVString>,
+    /// <https://notifications.spec.whatwg.org/#badge-url>
+    badge: Option<USVString>,
+    /// <https://notifications.spec.whatwg.org/#concept-language>
+    lang: DOMString,
+    /// <https://notifications.spec.whatwg.org/#silent-preference-flag>
+    silent: Option<bool>,
+    /// <https://notifications.spec.whatwg.org/#tag>
+    tag: DOMString,
+    /// <https://notifications.spec.whatwg.org/#concept-origin>
+    #[no_trace] // ImmutableOrigin is not traceable
+    origin: ImmutableOrigin,
+    /// <https://notifications.spec.whatwg.org/#vibration-pattern>
+    vibration_pattern: Vec<u32>,
+    /// <https://notifications.spec.whatwg.org/#timestamp>
+    timestamp: u64,
+    /// <https://notifications.spec.whatwg.org/#renotify-preference-flag>
+    renotify: bool,
+    /// <https://notifications.spec.whatwg.org/#require-interaction-preference-flag>
+    require_interaction: bool,
+    /// <https://notifications.spec.whatwg.org/#actions>
+    #[ignore_malloc_size_of = "NotificationAction"]
+    actions: Vec<NotificationAction>,
```

Comment:
This should not be `NotificationAction`, but rather struct not exposed to web content representing https://notifications.spec.whatwg.org/#action

See the public `actions` getter, which creates `NotificationAction` and returns them to web content. 

It may seem similar, but if the spec makes a difference, it's likely because there is a reason somewhere(or if not, you should file an issue to change the spec).